### ISSUE TITLE
Give a function to retry a Future

### ISSUE THREAD

**Issue by Oliboy50** [tokio-rs/tokio#5596]
## Problem

As a backend developer, I love the `tokio::time::timeout` function! 

But in backend development, we also often need a `retry` feature (which is also currently not given by the awesome `reqwest` crate for instance).

## Describe the solution you'd like

Some APIs like what the `tokio_retry` crates does.

Example from their README:
```rust
use tokio_retry::Retry;
use tokio_retry::strategy::{ExponentialBackoff, jitter};

async fn action() -> Result<u64, ()> {
    // do some real-world stuff here...
    Err(())
}

#[tokio::main]
async fn main() -> Result<(), ()> {
    let retry_strategy = ExponentialBackoff::from_millis(10)
        .map(jitter) // add jitter to delays
        .take(3);    // limit to 3 retries

    let result = Retry::spawn(retry_strategy, action).await?;

    Ok(())
}
```

Where `ExponentialBackoff` and `jitter` are really must-have for any resilient software, so the expected solution should expose these too... 

Note that I personally don't think that the current naming of functions are really goods (e.g. `Retry::spawn` feels odd), so it's up to the writer of this feature to choose better names if possible.

## Describe alternatives you've considered

Using `tokio_retry`... but it feels a bit unmaintained so I'd love seeing it in `tokio` directly (since there is a `timeout` function, as a Tokio user I also expect having a `retry` function).

## Example of what it may looks like to send a resilient request to another service

```rust
use tokio::retry::{retry, strategy::{ExponentialBackoff, jitter}};
use tokio::time::{Duration, timeout};

let retry_strategy = ExponentialBackoff::from_millis(10)
    .map(jitter) // add jitter to delays
    .take(3);    // limit to 3 retries
let request = timeout(Duration::from_millis(500), reqwest_client.get(url).send());

let Ok(response) = retry(retry_strategy, request).await else {
    tracing::Span::current().record("result", "timeout");
    return Err(RepositoryError::Timeout)
};
```

**Comment by Darksonn** [comment]
I don't think this is a good fit for the main Tokio crate. Maybe it could go in `tokio-util`, but I think the better outcome would be if `tokio-retry` became maintained.

**Comment by Oliboy50** [comment]
I also noticed that [a `tower::retry` module exists](https://docs.rs/tower/latest/tower/retry/index.html), but it probably does not fit for all Future types like `tokio::time::timeout` does, no?

**Comment by Darksonn** [comment]
That's correct.

**Comment by Xuanwo** [comment]
Please try [backon](https://crates.io/crates/backon) and let me know what you think! :laughing:

**Comment by Oliboy50** [comment]
@Xuanwo indeed, it seems like it would be [a good crate](https://docs.rs/backon/latest/backon/struct.ExponentialBuilder.html) for me to use as a substitute until Tokio supports retrying Futures, thanks 👍
